SMTChecker: Upgrade CVC4 to cvc5 and switch from API to SMT-LIB2 interface

blishko opened the pull request:
Instead of compiling `solc` itself with CVC4 support, it is now enough to have `cvc5` executable on PATH when running the compiler.
Instead of using API of CVC4, we now use SMT-LIB2 interface.
That means we write the queries to temporary SMT-LIB2 files and call the solver process directly to run on the file.

Depends on https://github.com/ethereum/solidity/pull/15102

Commit:
SMTChecker: Upgrade CVC4 to cvc5

Instead of compiling `solc` itself with CVC4 support, it is now enough
to have `cvc5` executable on PATH when running the compiler.
Instead of using API of CVC4, we now use SMT-LIB2 interface.
That means we write the queries to temporary SMT-LIB2 files and call
the solver process directly to run on the file.

Commit:
Update CI images

Commit:
SMTChecker: Allow using multiple external solvers in one analysis

We are using SMTCommand inside UniversalCallback to call external
solvers on queries produced my our engines.

Previous mechanism set the external solver once during initialization
and it was not possible to change it later. This meant, that it would
not be possible to use, e.g., Eldarica and cvc5 at the same time.

Here we move the proper setup for SMTCommand just before we call it.
This setup is customized by subclasses of (CHC)SmtLib2Interface, which
call corresponding external solvers.

ekpyron commented on CMakeLists.txt:
```diff
@@ -133,15 +133,15 @@ elseif (${Z3_FOUND})
-if (${CVC4_FOUND})
-  add_definitions(-DHAVE_CVC4)
-  message("CVC4 SMT solver found. This enables optional SMT checking with CVC4.")
+find_program(CVC5_PATH cvc5)
+if (CVC5_PATH)
+  add_compile_definitions(CVC5_PATH=${CVC5_PATH})
+  message("cvc5 SMT solver found. This enables optional SMT checking with cvc5.")
 endif()
```
I was about to say that this is bad, since if you then use ``CVC5_PATH`` somewhere, it'll get the path hardcoded - but you actually don't use it anywhere, right :-)? So the ``add_compile_definitions`` is superfluous anyways?

Should we even have any kind of ``cmake`` logic here? I've not looked at the rest of the PR yet, but I'd guess that CVC5 is purely a runtime dependency with the change, so we don't need to do anything special at cmake configure time, do we?

blishko replied:
That's a good point. I thought we need to give some sort of a message at configuration time, but most likely we don't.
Yes, `cvc5` is now a runtime dependency. It works the same way as `Eldarica`, so basically the executable must be present in PATH.

I'll remove this `cmake` logic.

blishko replied:
I am wondering about the message below, which warns if `Z3` nor `CVC4` (now `cvc5`) is not enabled/has not been found.

Should there be any information/warning at cmake configure time about available solvers?

nikola-matic commented on libsmtutil/CHCSmtLib2Interface.h:
```diff
@@ -79,17 +78,19 @@ class CHCSmtLib2Interface: public CHCSolverInterface
 	std::optional<smtutil::Expression> invariantsFromSolverResponse(std::string const& response) const;
 
+	/// Hook to setup external solver call
+	virtual void setupSmtCallback() {}
```
Should probably `smtAssert(false)` here.

blishko replied:
We cannot do that yet, because `--model-checker-solvers smtlib2` is a valid option now, and it will instantiate this base class.
What will happen is that it will try to call `SMTCommand`, but that will do nothing.

I think eventually we should just drop `smtlib2` as a possible choice for the solver (once also z3 migrates to the new usage).

nikola-matic commented on libsmtutil/SMTPortfolio.cpp:
```diff
@@ -32,26 +26,12 @@ using namespace solidity::frontend;
-	[[maybe_unused]] SMTSolverChoice _enabledSolvers,
-	std::optional<unsigned> _queryTimeout,
-	bool _printQuery
+	std::vector<std::unique_ptr<SolverInterface>> solvers,
```
```suggestion
	std::vector<std::unique_ptr<SolverInterface>> _solvers,
```

blishko replied:
Fixed.

nikola-matic commented on libsolidity/formal/EldaricaCHCSmtLib2Interface.h:
```diff
@@ -0,0 +1,41 @@
+protected:
+	void setupSmtCallback() override;
+
+	bool m_computeInvariants;
```
This can be private, no?

blishko replied:
Change to private both in `EldaricaCHCSmtLib2Inteface` and in `Cvc5SmtLib2Interface`.

nikola-matic reviewed (commented):
I've re-run the failing windows bytecode builds since they were failing (network err).

pgebal reviewed (commented):
Why don't we merge into develop?

There are still some references to CVC4 in `config.yml` and in some files `cmake/toolchains` dir. Search for `CVC4` in all files in the repo.

pgebal commented on libsolidity/formal/Cvc5SMTLib2Interface.cpp:
```diff
@@ -0,0 +1,36 @@
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libsolidity/formal/Cvc5SMTLib2Interface.h>
+
```
We usually don't put empty lines in between `<libsolidity...` includes

blishko replied:
Looking at a couple of files (BMC.*, CHC.*, Predicate.*), , we actually do separate includes from different directories. And always separating the main `.h` file for the current `.cpp` file from the other includes.

pgebal commented on libsolidity/formal/Cvc5SMTLib2Interface.cpp:
```diff
@@ -0,0 +1,36 @@
+	std::optional<unsigned int> _queryTimeout
+): SMTLib2Interface({}, std::move(_smtCallback), _queryTimeout)
+{
+
```
We could do without this empty line.

blishko replied:
Empty line removed!

pgebal commented on libsolidity/formal/EldaricaCHCSmtLib2Interface.cpp:
```diff
@@ -0,0 +1,37 @@
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libsolidity/formal/EldaricaCHCSmtLib2Interface.h>
+
```
Same as in CVC5 interface.

pgebal commented on libsolidity/interface/SMTSolverCommand.cpp:
```diff
@@ -51,6 +51,22 @@ void SMTSolverCommand::setEldarica(std::optional<unsigned int> timeoutInMillisec
+	else
+	{
+		m_arguments.push_back("--rlimit");
+		m_arguments.push_back(std::to_string(12000));
```
Optionally we could define 12000 as a named constant so it's not buried here.

blishko replied:
Where do you suggest to put the constant?

pgebal replied:
I admit I don't know what's the best place. First I thought about using `#define`, but I've learned it's not recommended. We can leave it like it is now.

pgebal reviewed (changes requested):
Very minor comments + let's update the changelog and remove references to CVC4 in cmake/toolchain and config.yml.

blishko commented:
> Very minor comments + let's update the changelog and remove references to CVC4 in cmake/toolchain and config.yml.

Added an entry to changelog and removed the missed references to CVC4. Thanks!

nikola-matic commented on Changelog.md:
```diff
@@ -6,7 +6,7 @@ Language Features:
  * EVM: Support for the EVM version "Prague".
  * SMTChecker: Add CHC engine check for underflow and overflow in unary minus operation.
-
+ * SMTChecker: Replace CVC4 as a possible BMC backend with cvc5.
```
```suggestion
 * SMTChecker: Replace CVC4 as a possible BMC backend with cvc5.
 
```

blishko replied:
Fixed!

r0qs reviewed (commented):
Looks good to me, but I believe we also need to update solc-js. For instance, see: https://github.com/ethereum/solc-js/blob/fa19cf85208d24f10e29646f6e33b1ce23719341/smtsolver.ts#L21

r0qs commented:
> Why don't we merge into develop?

Not sure if this was already answered, but the reason is because this PR relies on `cvc5` being installed on the images used by  the CI. Thus it depends on https://github.com/ethereum/solidity/pull/15102 being merged first.

r0qs commented on .circleci/config.yml:
```diff
@@ -1104,7 +1104,7 @@ jobs:
     environment:
       <<: *base_ubuntu2204_large_env
       CMAKE_BUILD_TYPE: Debug
-      CMAKE_OPTIONS: -DCMAKE_CXX_STANDARD=20 -DUSE_CVC4=OFF
```
Wait, doesn't this mean that `cvc5` will be used if found? Should this be the default behavior now in these images?

blishko replied:
`cvc5` will be used only in tests that explicitly request it.
It is now runtime dependency, not a compile-time dependency, as `CVC4` was.

Not sure what you mean by default behaviour. As I understand it, default behaviour does not change.

r0qs replied:
Oh, right. Previously, we explicitly disabled `cvc4` on those images, so I was concerned that this change might cause issues. However, everything seems to be fine indeed.

r0qs commented:
> > Looks good to me, but I believe we also need to update solc-js. For instance, see: https://github.com/ethereum/solc-js/blob/fa19cf85208d24f10e29646f6e33b1ce23719341/smtsolver.ts#L21
> 
> Hmm, OK, yes, we should probably do that. Even though, it is probably an independent thing, if I understand it correctly, no?

Yes, it is independent. I just wanted to highlight it here. It will just be a matter of changing the settings and the documentation. I tested solc-js with your branch, and it works fine. We might want to add some tests to solc-js for this, although we currently don't test Z3 there either.
